core/src/update.rs

```diff
@@ -0,0 +1,256 @@
+//! Trie update logic.
+//!
+//! The core of this module is the Multi-Update Algorithm, by which we can cause the same effect
+//! to a trie as a series of standalone put and delete operations, but with a minimal amount of
+//! hashing and revisiting of nodes.
+//!
+//! ## Multi-Update Algorithm
+//!
+//! The algorithm is based off of the observation that a set of put and delete operations can be
+//! partitioned into groups based on which terminal node their keys currently look up to.
+//! Each terminal node is then replaced with the sub-trie resulting from the set of given updates,
+//! and the trie is compacted into its smallest possible form, and hashed.
+//!
+//! For example,
+//!   - Replacing a single leaf node with another leaf node in the case of the previous leaf
+//!     being deleted and a new one with the same key or at least key prefix being put.
+//!   - Replacing a single leaf node with a terminator, in the case of deleting the leaf which was
+//!     there prior.
+//!   - Replacing a terminator with a leaf, in the case of a single put operation with that prefix
+//!   - Replacing a leaf node or terminator with a larger sub-trie in the case of multiple puts for
+//!     keys beginning with that prefix, possibly preserving the initial leaf.
+//!
+//! We refer to this as sub-trie replacement.
+//!
+//! Any newly created terminator nodes must be "compacted" upwards as long as their sibling is a
+//! terminator or a leaf node to create the most tractable representation. We combine this operation
+//! with hashing up towards the root, which will be described in a moment.
+//!
+//! Any changes in the trie must be reflected in the hashes of the nodes above them, all the way
+//! up to the root. When we replace a terminal node with a new sub-trie, we apply the compaction
+//! and hashing operations up to the point where no subsequently altered terminal will affect its
+//! result. The last terminal finishes hashing to the root. We refer to this as partial compaction.
+
+use crate::cursor::Cursor;
+use crate::trie::{self, KeyPath, NodeHasher, NodeHasherExt, NodeKind, ValueHash};
+
+use bitvec::prelude::*;
+
+/// Apply an update, represented as a sequence of operations, to the trie
+/// navigated by the given cursor.
+///
+/// The `ops` provided should be sorted ascending, lexicographically by key. This bails early when
+/// encountering unsorted keys.
+///
+/// Any existing terminal node that a key looks up to is implicitly deleted, unless its key and
```

Comment:
This API may seem weird at first and should be used with care, but it does align with our expectations for how Nomt will work under the hood: we need to pre-load the terminal for any written value in any case, and we should use that to craft the input to this algorithm.